### DISCUSSION TITLE
Can Jena convert a SPARQL endpoint to a Model?

### DISCUSSION THREAD

**Discussion by chenkuncloud** [apache/jena#1300] [Q&A] [answered]
Such as ```Fuseki's SPARQL endpoint```, can it be converted by ```Jena``` to a [Model](https://jena.apache.org/documentation/javadoc/jena/org/apache/jena/rdf/model/Model.html)?

This makes it very convenient to use the ```JAVA API``` to query.

If not for the time being, will it be considered in the future?

Best Regards

**Comment by afs** [comment]
Please review the previous discussions:

Issues https://github.com/apache/jena/issues/1243 and https://github.com/apache/jena/issues/1246

Networks are much slower than local method calls. In designing a system to work with local data and remote data, Jena has [RDFConnection](https://jena.apache.org/documentation/rdfconnection/) and other APIs. The fine grained model APIs is not the best starting point.

Specifically for arbitrary SPARQL endpoints using only SPARQL standard protocols, blank nodes will be problematic otherwise.

**Reply by Claudenw** [reply]
Years ago I worked on a project that did something similar.  We made remote endpoints appear as part of the graph behind Fuseki.  The techniques and results can be explored in the journal article: https://www.researchgate.net/publication/315120429_BioFed_Federated_query_processing_over_life_sciences_linked_open_data

But it still does not provide Model access.  Attempting to implement a Model that makes requests across the network will yield a Model that is so slow as to be unusable.

**Comment by namedgraph** [comment]
The DatasetAccessor interface is now deprecated, but it can be used to
abstract local and remote Model access.

We have a Service interface (and implementations) that can either be based
on a Dataset or a remote endpoints:
https://github.com/AtomGraph/Core/blob/master/src/main/java/com/atomgraph/core/model/Service.java

**Reply by afs** [reply]
That is like `RDFConnection` - fetch-modify-replace (SPARQL Graph Store Protocol) - which isn't what the question asks about.

**Comment by chenkuncloud** [comment]
Thank you!

I think I asked a very naive question, the graph is still at the remote endpoint, just like a full table query.

I know [RDFConnection.fetch()](https://jena.apache.org/documentation/javadoc/rdfconnection/org.apache.jena.rdfconnection/org/apache/jena/rdfconnection/RDFConnection.html) :
```
try (RDFConnection conn = RDFConnectionRemote.service(endpoint).build()) {
	Model model = conn.fetch(graphName);
	......
}
```
But if graph is very large, there will be errors: ```406 - Not Acceptable``` :
```
Exception in thread "main" org.apache.jena.atlas.web.HttpException: 406 - Not Acceptable
	at org.apache.jena.http.HttpLib.exception(HttpLib.java:231)
	at org.apache.jena.http.HttpLib.handleHttpStatusCode(HttpLib.java:161)
	at org.apache.jena.http.HttpRDF.execGetToInput(HttpRDF.java:165)
	at org.apache.jena.http.HttpRDF.httpGetToStream(HttpRDF.java:133)
	at org.apache.jena.http.HttpRDF.httpGetToStream(HttpRDF.java:128)
	at org.apache.jena.sparql.exec.http.GSP.GET(GSP.java:159)
	at org.apache.jena.rdflink.RDFLinkHTTP.get(RDFLinkHTTP.java:394)
	at org.apache.jena.rdflink.RDFConnectionAdapter.fetch(RDFConnectionAdapter.java:146)
	at com.test.Test.main(Test.java:21)
```

**Reply by afs** [reply]
`406 - Not Acceptable` is not necessarily related to graph size. It means the remote end rejected the request. Check the server logs for more details and also check the graph name.

If it is (very) large graph, use SPARQL to query and manipulate it.

**Comment by chenkuncloud** [comment]
But I wonder if a set of Fuseki special operations can be provided in the future? Like [Virtuoso](http://docs.openlinksw.com/virtuoso/rdfnativestorageprovidersjena/#rdfnativestorageprovidersjenaexamples3):
```
VirtGraph graph = new VirtGraph("xxx/sparql", "jdbc:virtuoso://xxx:1111", "dba", "xxx");
VirtModel model = new VirtModel(graph);
......
```

This will make Fuseki much convenient to operate.

Or that's not what Fuseki had in mind.

**Reply by afs** [reply]
> can be provided in the future

Apache Jena is an open source project. Contributions are welcome, including by Github pull requests.

It needs a private protocol to support blank node labels. The RDF Thrift format provides this feature.

It will be slow as the data scales (as the Virtuoso feature, and it is impractical for client-side SPARQL query).

`RDFConnection` is a better architecture to abstract away from whether the data is local or remote.

Issues https://github.com/apache/jena/issues/1243 and https://github.com/apache/jena/discussions/1246

**Reply by chenkuncloud** [reply]
Thank you. Looking forward to having that feature in the future!

**Comment by afs** [comment, marked as answer]
Answered by #1246.
